docs/DifferentiableFunctions.md

```diff
@@ -0,0 +1,444 @@
+    /// layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    /// - Returns: A tuple containing the output and the backpropagation function. The
+    ///   backpropagation function (a.k.a. backpropagator) takes a direction vector and returns the
+    ///   gradients at the layer and at the input, respectively.
+    func appliedForBackpropagation(to input: Input)
+        -> (output: Output,
+            backpropagator: (_ direction: Output.CotangentVector)
+                -> (layerGradient: CotangentVector, inputGradient: Input.CotangentVector)) {
+        let (out, pullback) = valueWithPullback(at: input) { layer, input in
+            return layer.applied(to: input)
+        }
+        return (out, pullback)
+    }
+}
+
+// Example neural network layer.
+struct DenseLayer: Layer {
+    var weight: Tensor<Float>
+    var bias: Tensor<Float>
+
+    @differentiable
+    func applied(to input: Tensor<Float>) -> Tensor<Float> {
+        return matmul(input, weight) + bias
+    }
+}
+
+// Example usage of `appliedForBackpropagation(to:)`.
+let dense = DenseLayer(weight: [[1, 1], [1, 1]], bias: [1, 1])
+let input: Tensor<Float> = [[3, 3]]
+let seed: Tensor<Float> = [[1, 1]]
+
+let (output, backprop) = dense.appliedForBackpropagation(to: input)
+let (𝛁dense, 𝛁input) = backprop(seed)
+
+dump(𝛁dense)
+// ▿ DenseLayer.AllDifferentiableVariables
+//   - weight: [[3.0, 3.0], [3.0, 3.0]]
+//   - bias: [1.0, 1.0]
+print(𝛁input)
+// [[2.0, 2.0]]
+```
+
+## Providing a custom derivative
+
+Use the `@differentiating` attribute to mark a function as a custom derivative
+for another function. This is useful for registering derivatives for primitive
+operations.
+
+Note: currently, the `@differentiating` attribute can only be used to define
+derivatives for functions in the same module. We plan to lift this limitation
+soon so that derivatives can be retroactively declared for functions in other
+modules - [see this forum
+discussion](https://forums.swift.org/t/help-needed-with-retroactive-differentiability/19927)
+for more information.
+
+```swift
+import Darwin
+
+func sillyExp(_ x: Float) -> Float {
+    let 𝑒 = Float(M_E)
+    print("Taking 𝑒(\(𝑒)) to the power of \(x)!")
+    return pow(𝑒, x)
+}
+
+@differentiating(sillyExp)
+func sillyDerivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
+    let y = sillyExp(x)
+    return (value: y, pullback: { _ in y })
+}
+
+print(gradient(of: sillyExp)(3))
+// Taking 𝑒(2.7182817) to the power of 3.0!
+// 20.085535
+```
+
+## Constructing a `@differentiable` function from a derivative
+
+Given a function and its derivative, it is possible to construct a
+`@differentiable` version of the function using the
+`differentiableFunction(from:)` helper function defined in the standard library.
+
+Here's an example:
+```swift
+let multiply: @differentiable (Float, Float) -> Float =
+    differentiableFunction(from: { x, y in (value: x * y, pullback: { v in v * y, v * x }))
```

Comment:
Syntax error on the pullback closure. Don’t think this will compile.

Reply:
Thanks for the catch! Fixed.